lang/src/org/partiql/lang/syntax/PartiQLParser.kt

```diff
@@ -0,0 +1,162 @@
+import org.antlr.v4.runtime.Token
+import org.antlr.v4.runtime.tree.ParseTree
+import org.partiql.lang.ast.ExprNode
+import org.partiql.lang.ast.toExprNode
+import org.partiql.lang.domains.PartiqlAst
+import org.partiql.lang.errors.ErrorCode
+import org.partiql.lang.errors.Property
+import org.partiql.lang.errors.PropertyValueMap
+import org.partiql.lang.types.CustomType
+import org.partiql.lang.util.getIonValue
+import org.partiql.lang.util.getPartiQLTokenType
+import org.partiql.lang.visitors.PartiQLVisitor
+import java.nio.charset.StandardCharsets
+import org.partiql.grammar.parser.generated.PartiQLParser as GeneratedParser
+import org.partiql.grammar.parser.generated.PartiQLTokens as GeneratedLexer
+
+/**
+ * Extends [Parser] to provide a mechanism to parse an input query string. It internally uses ANTLR's generated parser,
+ * [GeneratedParser] to create an ANTLR [ParseTree] from the input query. Then, it uses the configured [PartiQLVisitor]
+ * to convert the [ParseTree] into a [PartiqlAst.Statement].
+ */
+internal class PartiQLParser(
+    private val ion: IonSystem,
+    val customTypes: List<CustomType> = listOf()
+) : Parser {
+
+    override fun parseAstStatement(source: String): PartiqlAst.Statement {
+        // TODO: Research use-case of parameters and implementation -- see https://github.com/partiql/partiql-docs/issues/23
+        val parameterIndexes = getNumberOfParameters(source)
+        val lexer = getLexer(source)
+        val tree = try {
+            parseQuery(lexer)
+        } catch (e: StackOverflowError) {
+            val msg = "Input query too large. This error typically occurs when there are several nested " +
+                "expressions/predicates and can usually be fixed by simplifying expressions."
+            throw ParserException(msg, ErrorCode.PARSE_FAILED_STACK_OVERFLOW, cause = e)
+        }
+        val visitor = PartiQLVisitor(ion, customTypes, parameterIndexes)
+        return visitor.visit(tree) as PartiqlAst.Statement
+    }
+
+    private fun parseQuery(lexer: Lexer): ParseTree {
+        val parser = getParser(lexer)
+        return parser.statement()
+    }
+
+    private fun getLexer(source: String): Lexer {
+        val inputStream = CharStreams.fromStream(source.byteInputStream(StandardCharsets.UTF_8), StandardCharsets.UTF_8)
+        val lexer = GeneratedLexer(inputStream)
+        val handler = TokenizeErrorListener(ion)
+        lexer.removeErrorListeners()
+        lexer.addErrorListener(handler)
+        return lexer
+    }
+
+    private fun getParser(lexer: Lexer): GeneratedParser {
+        val tokens = CommonTokenStream(lexer)
+        val parser = GeneratedParser(tokens)
+        val handler = ParseErrorListener(ion)
+        parser.removeErrorListeners()
+        parser.addErrorListener(handler)
+        return parser
+    }
+
+    /**
+     * Create a map where the key is the index of a '?' token relative to all tokens, and the value is the index of a
```

Comment:
Probably a good idea to discuss 'why' this is needed.

I.e., the visitor visits the params in 2-dimensional tree traversal order, but the ordinal of the parameters is implicit in the 1-dimensional traversal of the token stream.


It might also be possibly to "wrap" the `GeneratedLexer` in your own `TokenSource`-derivative that delegates to to all other methods, but "wraps" `nextToken` to count tokens as they are emitted by the Lexer. The downside is more complicated lifetime of shared objects, but it would only require a single lexer pass through the input.